paddle/fluid/compiler/piano/shape.cc

```diff
@@ -0,0 +1,103 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/fluid/compiler/piano/shape.h"
+#include <algorithm>
+#include <utility>
+#include "paddle/fluid/compiler/piano/layout.h"
+#include "paddle/fluid/string/string_helper.h"
+
+namespace paddle {
+namespace piano {
+
+Shape::Shape(const note::ShapeProto& proto) {
+  element_type_ = proto.element_type();
+  dimensions_.insert(dimensions_.end(), proto.dimensions().begin(),
+                     proto.dimensions().end());
+  if (proto.has_layout()) {
+    layout_ = Layout(proto.layout());
+  }
+}
+
+note::ShapeProto Shape::ToProto() const {
+  note::ShapeProto proto;
+  proto.set_element_type(element_type());
+  proto.mutable_dimensions()->Reserve(dimensions().size());
+  for (auto&& dim : dimensions()) {
+    proto.add_dimensions(dim);
+  }
+  if (has_layout()) {
+    *proto.mutable_layout() = layout().ToProto();
+  }
+
+  return proto;
+}
+
+std::string Shape::ToString() const {
+  auto dtype_name = note::ElementType_Name(element_type());
+  std::transform(dtype_name.begin(), dtype_name.end(), dtype_name.begin(),
+                 [](auto&& c) { return std::tolower(c); });
+
+  std::vector<std::string> dim_names;
+  std::transform(dimensions().begin(), dimensions().end(),
+                 std::back_inserter(dim_names),
+                 [](auto&& dim) { return std::to_string(dim); });
+
+  return paddle::string::format_string(
+      "%s[%s]%s", dtype_name.c_str(),
+      paddle::string::join_strings(dim_names, ',').c_str(),
+      layout().ToString().c_str());
+}
+
+Signature::Signature(const note::SignatureProto& proto) {
+  std::transform(proto.parameters().begin(), proto.parameters().end(),
+                 std::back_inserter(parameters_),
+                 [](auto&& shape_proto) { return Shape(shape_proto); });
+  *mutable_result() = Shape(proto.result());
+  parameter_names_.insert(parameter_names_.end(),
+                          proto.parameter_names().begin(),
+                          proto.parameter_names().end());
+}
+
+note::SignatureProto Signature::ToProto() const {
+  note::SignatureProto proto;
+  proto.mutable_parameters()->Reserve(parameters().size());
+  for (auto&& shape : parameters()) {
```

Comment:
Same question: can we use `const auto &` ?

Reply:
done